### ISSUE TITLE
Web-sockets upgrade not working in alphav3

### ISSUE THREAD

**Issue by shashankn91** [istio/istio#11498]
**Describe the bug**
Prefix matching rules to upgrade the connection to web sockets are applied(kubctl apply) . When a request is made to WebSocket URL 404 occurs . It was observed that the rules are not seen in the envoy routes(istioctl proxy-config routes).

**Expected behavior**
For URI's filtered Web sockets must be working.

**Steps to reproduce the bug**
apply the Virtual Service rules and you see envoy routes not being updated. 

```
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: fashtagbackend-virtual-service
  namespace: fashtagbackend
spec:
  gateways:
  - fashtagbackend-gateway
  hosts:
  - 'fashtagbackend.fashtag.info'
  - fashtagbackend-service.backend.svc.cluster.local
  http:
  - match:
    - uri:
        prefix: "/replenishment/"
    route:
    - destination:
        host: fashtagbackend-service
      weight: 100
    websocketUpgrade: true
  - match:
    - uri:
        prefix: "/inventoryCount/"
    route:
    - destination:
        host: fashtagbackend-service
      weight: 100
    websocketUpgrade: true
  - match:
    - uri:
        prefix: "/transfer"
    route:
    - destination:
        host: fashtagbackend-service
      weight: 100
    websocketUpgrade: true
  - match:
    - uri:
        prefix: "/api/"
    route:
    - destination:
        host: fashtagbackend-service
      weight: 100
```

**Version**
kubectl version - Major:"1", Minor:"13"
istioctl version - 1.0.5
**Installation**
Istio is installed using helm
**Environment**
AWS

**Comment by ymesika** [comment]
Websocket upgrade should be enabled by default.
Can you try the same configuration after removing the `websocketUpgrade: true` entries from the routes?

**Comment by shashankn91** [comment]
I did remove the websocketUpgrade:true . I am still having the same issue.
Below is the configuration for route.
```
  http:
  - route:
    - destination:
        host: fashtagbackend-service


```

Registration of web sockets URL

```
INFO  [2019-02-04 09:08:43,144] io.dropwizard.websockets.WebsocketBundle: Registering websocket endpoints: 

    WS      /replenishment/{replnishmentRequestId} (com.fashtag.sockets.inventory.replenishment.ReplenishmentServer)

```

Log which says 404 when open socket connetion request is sent.
`27.59.80.71 - - [04/Feb/2019:09:15:56 +0000] "GET /replenishment/33b4a605-c72d-4837-8df4-ab2d5f5c807e HTTP/1.1" 404 282 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36" 3
`

**Comment by shashankn91** [comment]
Any suggestions or any updates?

**Comment by ymesika** [comment]
How can I reproduce it on my dev environment? Do you have some accessible alternative to your WSS service?

**Comment by Stono** [comment]
@ymesika just reading this, are you saying we don't need to specify the webSocketUpgrade value any more (i'm on 1.0.6)

**Comment by ymesika** [comment]
@Stono That's correct.
Since 1.0 we actually no longer take the VS field and always turn it on in the listener.

@rshriram should we remove the deprecated field from the VS def in 1.1? I guess it's too late.

**Comment by shashankn91** [comment]
@ymesika webSocketUpgrade value should be given and it is working with the upgraded version (1.1)of Istio. Maybe it wasn't working because WebSockets was in beta in 1.0.5.

**Comment by Stono** [comment]
The field should definitely be removed if it is redundant now

**Comment by shashankn91** [comment]
@Stono I don't think it is redundant. There has to be information which can let Istio know to bypass the request. If there is no field to do the same request can go to any of the pods in that virtual service which is not desirable in WebSockets case.

**Comment by Stono** [comment]
I'm pretty sure that's what mandar is saying above, eg websocket upgrades are supported automatically on all routes.

Routing some requests to a different pod is a different concern

**Comment by Stono** [comment]
PS.  We removed all referenced to websocketUpgrade and websockets still work 😎

**Comment by shashankn91** [comment]
Let me give a try by removing websocketUpgrade . @rshriram I would be really happy to know how is Istio handling WebSockets without the websocketUpgrade flag.

**Comment by aaronjwood** [comment]
I am also seeing issues with websockets when the sidecar is in place (404's from Envoy). In my case I'm not using a virtual service. The service I have that listens for websocket connections is using `tcp` as the port name. Shouldn't that translate to L4 proxying and not care about what's on top?

**Comment by aaronjwood** [comment]
Looking at all the listeners for my pod I see that there are websocket upgrade types applied:
```
istioctl proxy-config listeners <my_pod> -n <my_ns> -o json | grep -i websocket

                                    "upgrade_type": "websocket"
                                    "upgrade_type": "websocket"
                                    "upgrade_type": "websocket"
                                    "upgrade_type": "websocket"
                                    "upgrade_type": "websocket"
                                    "upgrade_type": "websocket"
                                    "upgrade_type": "websocket"
                                    "upgrade_type": "websocket"
                                    "upgrade_type": "websocket"
                                    "upgrade_type": "websocket"
                                    "upgrade_type": "websocket"
                                    "upgrade_type": "websocket"
                                    "upgrade_type": "websocket"
                                    "upgrade_type": "websocket"
                                    "upgrade_type": "websocket"
                                    "upgrade_type": "websocket"
                                    "upgrade_type": "websocket"
```

Now for the two ports that the service listens on:
```
istioctl proxy-config listeners <my_pod> -n <my_ns>
...
100.96.1.16        60062     TCP
100.96.1.16        60061     HTTP
...
```

The port that the websocket server listens on (TCP) shows nothing:
```
istioctl proxy-config listeners <my_pod> -n <my_ns> --port 60062 -o json | grep -i websocket
```

For the other port that this service listens on (HTTP) I see the websocket upgrade types applied:
```
istioctl proxy-config listeners <my_pod> -n <my_ns> --port 60061 -o json | grep -i websocket
                                    "upgrade_type": "websocket"
                                    "upgrade_type": "websocket"
                                    "upgrade_type": "websocket"
``` 

This leads me to believe that my service should be using HTTP and not TCP. If I edit my service and change the TCP port to HTTP I see the listener take effect:
```
...
100.96.1.16        60062     HTTP
100.96.1.16        60061     HTTP
....
```

and the websocket upgrade types are applied but I still see the 404's in Envoy:
```
[2019-04-26T23:06:34.050Z] "GET /<my_endpoint> HTTP/1.1" 404 NR "-" 0 0 0 - "100.96.3.1" "Go-http-client/1.1" "d1513d85-3e0b-4399-ad7c-986c544f4b17" "<external_dns>:80" "-" - - 100.96.3.26:80 100.96.3.1:63790 -
```

**Comment by aaronjwood** [comment]
Scratch what I said above, I seem to have an even deeper issue. My virtual service config is not making it to Envoy...

**Comment by aaronjwood** [comment]
Well, I figured out why this was happening. I have two gateways and one of them is for this websocket thing. When I deleted the other gateway everything started flowing in. It seems that having multiple separate gateways poses a problem in this scenario. Is this something that's not supported anymore?

**Comment by sergeydeg** [comment]
Hi,  we have similar deployment as yours,  and all websocket connections working fine.
First thing I noticed,  if cluster contains multiple Gateway resources (we have one for each resource domain), then spec.servers.port.name must be unique across cluster. This requirements happened when we upgraded from 0.8 to 1.0.2.
Second,  istio-proxy (backed by Envoy)  is just reverse proxy,  and shall add Upgrade http header in each hop with initial connection handshake (by websocket protocol design). If your backend service implement http WS handshake in plain tcp, its ok,  but envoy doesn't know in this case,  that underground there is http proto, and doesn't make connection upgrade in this hop.
